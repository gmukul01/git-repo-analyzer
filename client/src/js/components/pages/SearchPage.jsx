import React, { Component, Fragment } from "react";
import ReactLoading from "react-loading";

import Header from "containers/Header";
import SearchForm from "components/search/SearchForm";
import SearchResult from "components/search/SearchResult";
import SearchHistory from "components/search/SearchHistory";

export default class SearchPage extends Component {
  componentDidMount() {
    this.props.history.length === 0 && this.props.fetchSearchHistory();
  }

  componentDidUpdate(prevProps, prevState, snapshot) {
    if (prevProps.searchResult.items && prevProps.searchResult.items.length > 0 && !this.props.searchResult.items) {
      this.props.fetchSearchHistory();
    }
  }
  render() {
    const { history, searchResult, isLoading, errorMessage, addToSearchHistory, fetchRepoDetails, fetchSearchHistory } = this.props;
    return (
      <Fragment>
        <Header />
        <section className="content">
          <SearchForm fetchRepoDetails={fetchRepoDetails} />
          {(() => {
            if (isLoading) {
              return <ReactLoading type="spinningBubbles" color="black" />;
            } else if (errorMessage) {
              return <p>{`Error : ${errorMessage}`}</p>;
            } else if (searchResult && searchResult.items && searchResult.items.length > 0) {
              return <SearchResult result={searchResult} addToSearchHistory={addToSearchHistory} />;
            }
            return <SearchHistory history={history} fetchSearchHistory={fetchSearchHistory} />;
          })()}
        </section>
      </Fragment>
    );
  }
}
