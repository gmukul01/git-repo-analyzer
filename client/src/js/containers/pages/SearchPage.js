import { connect } from "react-redux";

import SearchPage from "components/pages/SearchPage";
import { addToSearchHistory, fetchRepoDetails, fetchSearchHistory } from "actions/search";

const mapStateToProps = state => ({
  userName: state.user ? state.user.name : undefined,
  history: state.search.history,
  searchResult: state.search.result,
  isLoading: state.search.isLoading,
  errorMessage: state.search.errorMessage
});

export default connect(
  mapStateToProps,
  {
    addToSearchHistory,
    fetchSearchHistory,
    fetchRepoDetails
  }
)(SearchPage);
