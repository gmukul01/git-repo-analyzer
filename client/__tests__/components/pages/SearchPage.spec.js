import React from "react";
import { shallow } from "enzyme";
import ShallowRenderer from "react-test-renderer/shallow";

import SearchPage from "components/pages/SearchPage";

describe("SearchPage Component", () => {
  const initialProps = {
    history: ["dummyHistory"],
    searchResult: {
      items: ["dummyItems"]
    },
    addToSearchHistory: jest.fn(),
    fetchRepoDetails: jest.fn(),
    fetchSearchHistory: jest.fn()
  };

  it("should call mockFetchSearchHistory when history is empty on mount", () => {
    const props = {
        ...initialProps,
        history: [],
        fetchSearchHistory: jest.fn()
      },
      wrapper = shallow(<SearchPage {...props} />);
    expect(props.fetchSearchHistory.mock.calls.length).toEqual(1);
  });

  it("should not call mockFetchSearchHistory when history is already present on mount", () => {
    const props = {
        ...initialProps,
        fetchSearchHistory: jest.fn()
      },
      wrapper = shallow(<SearchPage {...initialProps} />);
    expect(props.fetchSearchHistory.mock.calls.length).toEqual(0);
  });

  it("should match snapshot with when isLoading is true", () => {
    const props = {
        ...initialProps,
        isLoading: true,
        errorMessage: false
      },
      renderer = new ShallowRenderer(),
      result = renderer.render(<SearchPage {...props} />);
    expect(result).toMatchSnapshot();
  });

  it("should match snapshot with when errorMessage is true and isLoading is false", () => {
    const props = {
        ...initialProps,
        isLoading: false,
        errorMessage: true
      },
      renderer = new ShallowRenderer(),
      result = renderer.render(<SearchPage {...props} />);
    expect(result).toMatchSnapshot();
  });

  it("should match snapshot with when search result is present", () => {
    const props = {
        ...initialProps,
        isLoading: false,
        errorMessage: false
      },
      renderer = new ShallowRenderer(),
      result = renderer.render(<SearchPage {...props} />);
    expect(result).toMatchSnapshot();
  });

  it("should match snapshot with when search history is present", () => {
    const props = {
        ...initialProps,
        isLoading: false,
        errorMessage: false,
        searchResult: undefined
      },
      renderer = new ShallowRenderer(),
      result = renderer.render(<SearchPage {...props} />);
    expect(result).toMatchSnapshot();
  });

  it("should match snapshot with when search history and result is not present", () => {
    const props = {
        ...initialProps,
        isLoading: false,
        errorMessage: false,
        searchResult: undefined,
        history: undefined
      },
      renderer = new ShallowRenderer(),
      result = renderer.render(<SearchPage {...props} />);
    expect(result).toMatchSnapshot();
  });
});
