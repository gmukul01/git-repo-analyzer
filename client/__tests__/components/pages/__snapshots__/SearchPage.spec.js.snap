// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchPage page  should match snapshot with when errorMessage is true and isLoading is false 1`] = `
<React.Fragment>
  <Connect(Header) />
  <section
    className="content"
  >
    <SearchForm
      fetchRepoDetails={[MockFunction]}
    />
    <p>
      Error : true
    </p>
  </section>
</React.Fragment>
`;

exports[`SearchPage page  should match snapshot with when isLoading is true 1`] = `
<React.Fragment>
  <Connect(Header) />
  <section
    className="content"
  >
    <SearchForm
      fetchRepoDetails={[MockFunction]}
    />
    <Loading
      color="black"
      delay={0}
      height={64}
      type="spinningBubbles"
      width={64}
    />
  </section>
</React.Fragment>
`;

exports[`SearchPage page  should match snapshot with when search history and result is not present 1`] = `
<React.Fragment>
  <Connect(Header) />
  <section
    className="content"
  >
    <SearchForm
      fetchRepoDetails={[MockFunction]}
    />
  </section>
</React.Fragment>
`;

exports[`SearchPage page  should match snapshot with when search history is present 1`] = `
<React.Fragment>
  <Connect(Header) />
  <section
    className="content"
  >
    <SearchForm
      fetchRepoDetails={[MockFunction]}
    />
    <SearchHistory
      fetchSearchHistory={[MockFunction]}
      history={
        Array [
          "dummyHistory",
        ]
      }
    />
  </section>
</React.Fragment>
`;

exports[`SearchPage page  should match snapshot with when search result is present 1`] = `
<React.Fragment>
  <Connect(Header) />
  <section
    className="content"
  >
    <SearchForm
      fetchRepoDetails={[MockFunction]}
    />
    <PureComponent
      addToSearchHistory={[MockFunction]}
      result={
        Object {
          "items": Array [
            "dummyItems",
          ],
        }
      }
    />
  </section>
</React.Fragment>
`;
